Uncertainty html rendering

sankha555 opened the pull request:
* [x] #133 
* [x] added HTML snippets to display uncertainty information in HTML report
* [x] tests passed

sankha555 commented:
@matthiaskoenig I have added some snippet strings that can be rendered into HTML code fragments for the parameters of the uncertainties, in ```info_sbase()```. I have then changed the macros for the Jinja template in ```macros_sbml.html``` file to render this added information for the uncertainties. However I am a bit in doubt if the display of the uncertainties should be like what I have tried in this patch. I tested a prototype display using the ```uncertainty.xml``` example. I have attached the rendered HTML snapshot below for your reference. 

![image](https://user-images.githubusercontent.com/54751139/109217315-b39ebd00-77db-11eb-8b1c-a9fb9829a8cc.png)


As of now,  i have added this for the Parameters component only, if you think that this would be a correct approach, I can then do the same thing for all the other components which include the"distrib" plugin (e.g. parameters, initial assignements, events, etc). Also, I would add the same feature for the uncertainties as well, which would then enable creating the HTML fragments for the uncertainties in the ```info_sbase()``` itself without the need to do it in the macros file (which I suppose is the standard used for all the other components).

Please let me know your feedback about the additions and please let me know if I should make any changes to improve it. I will keep on adding the changes in the commits for your reviews, and in the end I would squash them all into one commit. Thanks!

sankha555 commented:
@matthiaskoenig Sorry that I could find time only during the late hours today to resume work on this issue. Although I haven't made any major changes, I have made a few observations regarding the rendering. I have extended the macros components for the EventAssignment, InitialAssignment, Rule components in the manner similar to Parameters to display the uncertainties info. 

I wanted to ask one thing that is it considered bad practice to manipulate the template tagging in the macros extensively rather than creating all the HTML snippets in the form of python strings and then directly using them in the template? Like the method I have used to template tag every attribute for the uncertainties info in the macros, is it an unsuitable approach? I had today tried to improve the uncertainties in form of strings in ```info_sbase()``` itself (I have left an instance of the ```u_str``` variable for your reference) but it was posing a display error due to which the entire alignement of the information was getting distorted. This was particularly arising also due to the ```id_html``` attribute of the uncertainties sbase. Hence, I have let the the template tagging stay as is in the macros. I will try again to resolve this issue of alignment by trying again with some CSS classes present in the project.

Please let me know if you have any more inputs for me on the work till now. I will continue to resolve the above issues by implementing your suggestions and exploring on my own. Thanks!

matthiaskoenig commented:
@sankha555 No worries about the exact rendering of the HTML. An important part of the GSOC rendering will be to update the HTML rendering. What is done now by Jinja will be part of Vue.js. I.e. the info dictionaries will not contain any HTML rendering information any more, but all the HTML rendering will happen in Vue.js components. 
For now this was mainly for you to have worked on the complete process of updating information and get it rendered in the current report.

matthiaskoenig commented on src/sbmlutils/resources/templates/macros_sbml.html:
```diff
@@ -245,6 +246,24 @@
 
 	<td>{{ item.sbo }}</td>
 	<td>{{ item.cvterm }}</td>
+    <td>
```
This is again too much code duplication!
Just define once a macro for the uncertainties and then reuse it on all objects!
See the `{{ id_info(item) }}` macro for an example.
It is defined once at the top and then reused in the other macros. Do the same strategy here instead of copying the same block everywhere.

matthiaskoenig commented on src/sbmlutils/resources/templates/macros_sbml.html:
```diff
@@ -270,6 +290,24 @@
 	<td>{{ item.derived_units }}</td>
 	<td>{{ item.sbo }}</td>
 	<td>{{ item.cvterm }}</td>
+    <td>
```
see comment above: code duplication

matthiaskoenig reviewed (changes requested):
Create macro for uncertainties and reuse it for all objects.

sankha555 commented:
@matthiaskoenig I just made the changes, please take a look at it. The possibility for the code re-use like the ```id_info``` macro completely slipped my mind, sorry for that. I have fixed the code duplication using the ```uncertainty_sbml``` macro. 
Also I have rendered the definition urls as anchor tag links like the other links in the reports. 

There was one issue with the math rendering for the math element of the uncertainty parameters, which I then rendered using the ```math()``` function. But inside ```math()```, it was extracting the ASTNode component of each item that was passed to it by default because by definition we are passing SBase items to ```math()```. This was raising an error  with the math elements of the parameters as they were already ASTNodes. So I added a check in the ```math()``` to see if the item was already an ASTNode, then skip the extraction of the ASTNode else extract it using ```item.getMath()```. Would this change be suitable Sir, or should I undo it? I had also tested it to be certain that it wasn't introducing new breaks with other models and it passed all the tests.

I understand a bit now that what has to be done in the GSoC project with the rendering after your explanation. I suppose that almost all the formatting and rendering will be done right on the frontend using Vue.js that time, unlike how we do in the Python files currently. This exercise gave me a good understanding of introducing new information dicts right from scratch and rendering them accordingly and has improved my understanding of the project even more now. Please let me know if I should make any more changes to it, I will do any required changes and upload for your reviews. Thanks!

matthiaskoenig commented:
Looks great. Merging this now and making a release.
